### ISSUE TITLE
optimize non-huge update processing

### ISSUE THREAD

**Issue by patrickzurek** [eXtensibleCatalog/xcmetadataservicestoolkit#631]
JIRA issue created by: banderson
Originally opened: 2011-06-23 07:14 PM

Issue body:
 (nt)

**Comment by patrickzurek** [comment]
JIRA Coment by user: banderson
JIRA Timestamp: 2011-06-23 07:14 PM

Comment body:

<p>
	we currently have to modes of processing records through the MST</p>
<p>
	1) optimized</p>
<p>
	2) non-optimized</p>
<p>
	Currently, the only actual difference in these modes is whether or not the tables have indexes on them.&nbsp; In the initial load, the indexes are not there.&nbsp; It isn&#39;t until after the first load that the indexes are created.&nbsp; So, when an update happens it starts in non-optimized mode.&nbsp; After X (100,000) number of records, it is determined that the MST should go into optimized mode.&nbsp; To do this, it drops the indexes, continues processing and adds the indexes back after processing completes.</p>
<p>
	The other aspect which I&#39;m proposing we change is the in memory caches.&nbsp; Currently when a harvest or service starts the MST loads certain data structures in full into memory so that it doesn&#39;t have to do one-off db queries.&nbsp; IOW, it starts in optimized mode.&nbsp; I propose that we change it so that it starts in non-optimized mode (one-off db queries) until it hits the threshhold.</p>
<p>
	This matters because doing non-huge update processing takes about 1000 longer than it needs to right now.</p>

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-01-27 08:19 PM

Comment body:

Running my Randy30 test today (on a server that has several million source records in the index and DB, a "normal" library load), which has updates for less than 30 source records, requires processing that seems to take large amounts of time 30-60 minutes (I didn't time this).

Ralph, for you MST in daily mode, do you have any idea how long the entire "update" process is taking?

John/Chris....do you have ideas on improving this?  I know Chris has expressed interest in making sure the daily processing is done in a reasonable time amount.

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-04-05 03:46 PM

Comment body:

Chris, I have moved this to 1.4 as good work that impacts CARLI while we wait for the 1.5 specs to be written.   What would it take to implement what Ben proposed?

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-04-23 01:25 PM

Comment body:

I was able to separate / encapsulate the use if the record cache during the update process. I can now easily switch between using a cache vs direct database calls. Now the question becomes "when should the MST decide to choose which method to use?". I will most likely be keying off the completeListSize attribute from the resumptionToken because I need to make this decision early on to be most effective. This attribute is optional, but since we do use it we can be confident it will be set. I suppose we could also decide to always begin without the cache and when we hit a certain threshold switch to a cache. What should we set for this threshold? Should we use a static number like 10000?  Should we base it on a percentage of the total repo size? What do you think?

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-05-04 03:36 PM

Comment body:

John,

I'm finding out that I need to also do away with caching unnecessary objects during updates in the case of a relatively small update (MST-277) because the caching process takes a LONG time to do (much longer than the actual updates would take).  So, basically, if we are dealing with a multimillion record repo, the updates will always take long even if you only update one (or zero!) record.

I've already pulled out the caching code in HarvestManager.  This solves the issue for the initial repo.  However, there is a similar caching process occurring in the GenericMetadataService object.  Basically for all services -- e.g., marcnormalization, marctoxctransformation -- the same thing we do vis a vis OAI IDs in HarvestManager, we do for previous statuses in Services.  I was about to follow the same pattern of encapsulating the caching process so I can make it an optional thing (i.e., use direct database calls instead of caching everything up front).  HOWEVER, the "fix" for GC 316 basically kills that idea.  Here's the code:

http://code.google.com/p/xcmetadataservicestoolkit/source/browse/trunk/mst-common/src/java/xc/mst/services/GenericMetadataService.java#479

The code persists ALL of the previous statuses for ALL the records in the service's repo.  Is this really necessary?  Would it be ok -- and still "fix" the issue in GC 316 -- if we were to only persist the records which have actually changed?  Because if all statuses need to be persisted, then I can not ever do away with the slow caching process (because we'd be going through all the records anyway, pulling out the cache wouldn't save us anything).

I am asking you this question because I currently do not know enough about how previous statuses affect the overall system (and GC 316 in particular).

Thanks,
Chris

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-05-08 12:07 PM

Comment body:

I solved some of the issues (trunk, rev 3231).  In particular, I sped up the processing of the repo harvesting.  There is still room for improvement in processing the MST services (GenericMetadataService code).  I will leave this open until I or someone else in the future understands the "previous statuses" concept enough to proceed.  Currently the MST retrieves all of the previous status information before it processes its services; it then persists this information at the end.  I currently do not know if retrieving and persisting this information for ALL of the records is necessary or not; if it isn't necessary, we could speed up "small" updates considerably if we only retrieved and persisted previous status information for records that actually get updated.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-05-08 06:02 PM

Comment body:

Guess what?  All this worry about speeding up these MST processes may perhaps be a drop in the bucket when compared to Solr indexing:

08 May 2012 14:47:30,357 xc.mst.utils.LogWriter:65  INFO [SolrWorkDelegate] - Start optimizing Solr index
08 May 2012 15:17:06,015 xc.mst.utils.LogWriter:65  INFO [SolrWorkDelegate] - Finished optimizing Solr index
08 May 2012 15:17:32,685 xc.mst.utils.LogWriter:65  INFO [SolrWorkDelegate] - Start optimizing Solr index
08 May 2012 15:50:06,821 xc.mst.utils.LogWriter:65  INFO [SolrWorkDelegate] - Finished optimizing Solr index
08 May 2012 15:58:52,328 xc.mst.utils.LogWriter:65  INFO [SolrWorkDelegate] - Start optimizing Solr index
08 May 2012 16:26:43,622 xc.mst.utils.LogWriter:65  INFO [SolrWorkDelegate] - Finished optimizing Solr index

The above is the final solr indexing for a repo consisting of ~12 million records and a small update of 8 records.  The update process officially began at 14:27.  So, do we really care if we were able to tighten the bolts on the first twenty minutes knowing full well that Solr indexing will take another 90-100 minutes?  If we run into an issue with update times taking too long, I think we'd need to focus more on the elephant in the room ;-)

For the record: I'm glad I was able optimize the could I was able to (along with clean up some of it, too).  I just don't think it's worth spending any more time working on this issue.  At least not at this juncture...  Chris

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-05-08 07:55 PM

Comment body:

John just reminded me that Solr indexing occurs during idle periods. Excellent point. I will leave this issue Incomplete, but still open for future enhancement.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2012-05-09 09:30 AM

Comment body:

[~cdelis] is right though - there is plenty of room left for improvement.

1)  the optimizing runs multiple times, at the end of each service processing, the indexer runs, and at the end of each of those times, it optimizes.  Once, at the very end, is desired. In my time working on it I had not figured out how to do that.

2) contrary to my comment to Chris, I believe the optimization is a black box solr call that I don't know how to interrupt.  Maybe it is possible but for now, one you start you will continue with that time consuming process.

3) the solr optimization thing should be split into a separate issue - there may already even be one.  MST-390 covers indexing speed - but I don't think that is the problem, it is the optimization (which we never bothered to do till the new indexing scheme was implemented).  MST-346 is related.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-05-09 02:32 PM

Comment body:

I think I finally "get it," re: previous statuses.  I found a way to do without cacheing the whole lot of records' previous statuses.  I am basically following the same pattern I used in saving process time in the Repo code; now we can achieve the savings multiple times (one for each MST service).  I will post the actual time savings as soon as I finish running tests.  --Chris

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-05-09 04:52 PM

Comment body:

For a repo containing ~6 million bibs and ~7 million holding records (yielding a total of ~13 normalized records and ~21 million XC schema records), I ran an update of 3 bibs and 5 holdings records (total of 8 records).  With the original code (pre-MST-277 code changes) the MST took 20 mins 00 secs to harvest, and process (via marcnormalization and marctoxctransformation).  After optimizing (doing away with the unnecessary caching of objects, OAI IDs), the time went down to 16 mins 07 secs.  And after optimizing the services code in GenericMetadataServices object (doing away with the unnecessary caching of objects, record IDs and previous statuses), the went down further to 10 mins 18 secs.  So in total we shortened the processing time 50%.  FYI, the GC 316 changes (Kyushu changes) was a red herring as it did NOT impact this issue, nor should these changes affect its functionality. I believe this work was worth doing and will now consider it completed. In trunk, svn rev. 3238  --Chris

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2012-05-09 04:58 PM

Comment body:

[~cdelis] Is most of the rest of the time the solr.optimize?

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-05-09 05:01 PM

Comment body:

I didn't factor solr optimize in my times above, since it's a separate issue.  The solr optimize still takes the same amount of time (90-100 minutes).

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-05-09 05:02 PM

Comment body:

I should NOT have used the word "optimizing" in my description, since it can lead to confusion (i.e., I did not mean SOLR optimization ;-)  Sorry for the confusion.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-05-10 04:46 PM

Comment body:

I am currently in the middle of running a "large" small update without caching OAI IDs, and it is S-L-O-W.  Wow, I had no idea it would be that slow.  It is so bad in my opinion that I'm going to almost completely do away with the non-caching option and stick with what we had in the beginning.  I am going to remove the option in the install.properties file and set the "large" update threshold from 10000 to something tiny like 100.  It really is that bad (I don't even wish to wait for my test run to complete ;-)  The code changes I made are OK, however.  Seems a bit cleaner, so all is not lost.  I was hoping for better results, though, for sure.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-05-10 04:49 PM

Comment body:

I forgot to point out in my previous comment that my test run consisted of a 9,999 record update: 5,000 bibs and 4,999 holdings.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-05-10 05:03 PM

Comment body:

The reverted changes are in trunk, rev. 3244

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-05-11 05:31 PM

Comment body:

I'm not giving up yet; I noticed that the mysql tables don't index oai_id which is a crucial for performance in this case.  I'm going to create the index and see if it makes a difference.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-05-16 03:53 PM

Comment body:

In trunk, rev. 3248
I added a new index on oai_id column in [repo].record_oai_ids
Re-added the install.properties property:

# Number of records considered to be a "large" harvest
# (Do not set this too small because it may degrade performance for smaller harvests.)
#LargeHarvestThreshold=10000

Cleaned up code a bit from last MST-277 commit.  Fixed bug where it didn't process the LargeHarvestThreshold properly

I believe MST-277 can be considered completely and officially "done."

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2012-05-22 09:30 AM

Comment body:

[~cdelis]Is the parameter in default.properties as well as install.properties?   The way it works is all properties go into default.properties with their default value, then they can be overridden in install.properties.  Is that how you implemented?

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2012-05-22 09:33 AM

Comment body:

The default  is set in code.  Thanks for pointing me to default.properties. I will set it there too.

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2013-01-30 01:54 PM

Comment body:

[~cdelis] can you give me a % improvement?  I would like to make s statement about this in the Mellong report.

**Comment by patrickzurek** [comment]
JIRA Comment by user: Chris Delis (cedelis)
JIRA Timestamp: 2013-01-30 05:13 PM

Comment body:

There isn't a fixed % I can give you, since it is not a fixed equation (very much depends on size of T0 and T1, mostly T0). I can run a sample set of some arbitrary # of T0 records, then run force T1 to cache all current records (via setting the LargeHarvestThreshold to a tiny number, i.e., < # T1 records), which is the "old" way of doing things -- vs. -- *not* cacheing, which is the "new" more efficient method.

Or... you can just use the historical % improvement I reported back on May 9th, 2012, which was 50%. Probably just as good of an estimation as any (even though it's a bit out of date, it still captures the basic idea of "let's *not* cache all the record states beforehand, but on a need-to-know basis").

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2013-01-30 05:32 PM

Comment body:

Capturing the following from a skype interaction.

rcook:  on the 277 issue...i saw your 50% estimate, but then I saw later comments saying you were revert back to older code, so I didn't know what to make of that.

cdelis: I reverted it back, before I re-instated it. The gains -- e.g., 50% -- didn't scale because a mysql table was lacking an index (i.e., the greater the # of recs, the progressively slower it got). Once I created the index, the mysql slowness disappeared, and the gains remained.
